app/grandchallenge/emails/emails.py

```diff
@@ -1,34 +1,111 @@
 from django.template.loader import render_to_string
-from django.utils.html import format_html
+from django.utils.html import format_html, strip_tags
 
+from grandchallenge.profiles.models import EmailSubscriptionTypes
+from grandchallenge.subdomains.utils import reverse
 
-def send_standard_email(*, site, subject, message, recipient, unsubscribable):
-    if not recipient.is_active:
-        # Do not send emails to inactive users
-        return
 
-    if (
-        unsubscribable
-        and not recipient.user_profile.receive_notification_emails
-    ):
-        # Do not send unsubscribable emails to users who have opted out
-        return
+def filter_recipients(recipients, subscription_type=None):
+    filtered_recipients = []
+    for recipient in recipients:
+        if not recipient.is_active:
+            # Do not send emails to inactive users
+            continue
 
-    send_mail(
+        if (
+            subscription_type == EmailSubscriptionTypes.NOTIFICATION
+            and not recipient.user_profile.receive_notification_emails
+        ) or (
+            subscription_type == EmailSubscriptionTypes.NEWSLETTER
+            and not recipient.user_profile.receive_newsletter
+        ):
+            # Do not send emails to users who have opted out
+            continue
+
+        filtered_recipients.append(recipient)
+
+    return filtered_recipients
+
+
+def send_standard_email_batch(
+    *, subject, message, recipients, subscription_type=None
+):
+    connection = get_connection()
+    messages = []
+    recipients = filter_recipients(
+        recipients=recipients, subscription_type=subscription_type
+    )
+    for recipient in recipients:
+        messages.append(
+            create_email_object(
+                recipient=recipient,
+                subject=subject,
+                message=message,
+                subscription_type=subscription_type,
+                connection=connection,
+            )
+        )
+    return connection.send_messages(messages)
+
+
+def get_unsubscribe_link(recipient, subscription_type):
+    if subscription_type == EmailSubscriptionTypes.NEWSLETTER:
+        return reverse(
+            "newsletter-unsubscribe",
+            kwargs={"token": recipient.user_profile.unsubscribe_token},
+        )
+    elif subscription_type == EmailSubscriptionTypes.NOTIFICATION:
+        return reverse(
+            "notification-unsubscribe",
+            kwargs={"token": recipient.user_profile.unsubscribe_token},
+        )
+    else:
+        return None
+
+
+def get_headers(unsubscribe_link):
+    if unsubscribe_link:
+        return {
+            "List-Unsubscribe": unsubscribe_link,
+            "List-Unsubscribe-Post": "List-Unsubscribe=One-Click",
+        }
+    else:
+        return None
+
+
+def create_email_object(
+    recipient, subject, message, connection, subscription_type=None
+):
+    site = Site.objects.get_current()
```

Comment:
By moving this here I think that you now do one query per email sent. This introduces an n+1 problem, which will be an issue when sending out the mailing list emails.

Reply:
Oh wow, you're absolutely right!